### ISSUE TITLE
Библиотека enb-ng-techs

### ISSUE THREAD

**Issue by Guria** [bem/bem-forum-content-ru#414]
У меня есть пара снипетов, которые потенциально оформить как библиотеку технологий для enb. Однако подобного опыта пока не имею.
```js
var path = require('path');

module.exports = require('enb/lib/build-flow').create()
    .name('angular-templates')
    .useFileList(['tmpl.html'])
    .target('target', 'templates.html')
    .justJoinFiles(function (filename, data) {
        return '<script type="text/ng-template" id="' + filename.split(path.sep).pop() + '">\n' + data + '\n</script>';
    })
    .createTech();
```
```js
var ngAnnotate = require("ng-annotate");
module.exports = require('enb/lib/build-flow').create()
    .name('ng-annotate')
    .defineRequiredOption('source')
    .defineOption('target')
    .target('target', '?.js')
    .useSourceText('source')
    .builder(function(source){
        return ngAnnotate(source, { add : true }).src;
    })
    .createTech();

```
Что скажете? Надо ли это оформлять как библиотеку?
@blond, поможешь?

**Comment by blond** [comment]
Конечно оформляй в библиотеку :)

Если интересно как написать тесты — обращайся ;)

**Comment by Guria** [comment]
К себе пока или в инкубатор сразу стоит?

**Comment by tadatuta** [comment]
@Guria досыпал доступов, так что можешь публиковать сразу в инкубатор

**Comment by Guria** [comment]
[done](https://github.com/bem-incubator/enb-ng-techs)
@blond Какую лучше версию enb в зависимости писать? За помощь с тестами буду благодарен.

**Comment by blond** [comment]
@Guria :+1: 

> Какую лучше версию enb в зависимости писать?

Последняя норм.

> За помощь с тестами буду благодарен.

Можно посмотреть на тесты в [enb-bem-techs](https://github.com/enb-bem/enb-bem-techs/tree/master/test/techs) или [enb-bh](https://github.com/enb-bem/enb-bh/blob/master/test/techs/bh-commonjs.test.js).

Как писать тесты:
1. Подменяем файловую систему с помощью [mock-fs](https://github.com/tschaub/mock-fs) для того, чтобы не работать с реальной.
2. Используем мок ENB ноды [https://github.com/enb-make/enb/blob/master/lib/test/mocks/test-node.js].
3. Для [технологий ожидающих список файлов](https://github.com/bem-incubator/enb-ng-techs/blob/master/techs/ng-templates.js#L46) используем [FileList](https://github.com/enb-make/enb/blob/master/lib/file-list.js), пример: https://github.com/enb-bem/enb-bh/blob/master/test/techs/bh-commonjs.test.js#L339-L342

**Comment by Guria** [comment]
Я ещё не приступал к тестам. Правильно ли я понимаю, что теперь ещё следует смотреть в сторону [mock-enb](https://github.com/enb-make/mock-enb) ?

**Comment by blond** [comment]
@Guria, да, всё так. Там по смыслу, пока, ничего не изменилось, просто код вынесен в отдельный модуль и обновлены зависимости.

**Comment by Guria** [comment]
@blond [Накидал](https://github.com/bem-incubator/enb-ng-techs/tree/tests) минимальные заглушки для тестов.
И прежде чем пытаться что-то тестировать решил сначала определить, что именно я буду тестировать.
Если для `ng-templates` я придумал целых 2 теста, то с `ng-annotate` сложнее. Проверять работу подключаемой библиотеки? Или просто надо как-то убедиться что библиотека была использована?

Ещё, кажется, я не очень удачно назвал свои технологии. В частности имя моей технологии совпадает с именем подключаемой библиотеки.

**Comment by blond** [comment]
> Проверять работу подключаемой библиотеки? Или просто надо как-то убедиться что библиотека была использована?

Если библиотека стабильна и код-обёртка просто использует её API, то я обычно проверяю только то, что это API выполняется. То, что результат должен быть правильным — ответственность тестов самой библиотеки.

Но если даже в патчевых версиях иногда что-нибудь ломается или за результат отвечает не только сама библиотека, но и код технологии — я стараюсь тестировать и то, что результат в итоге оказывается правильным. 

Пример второго — `enb-stylus`. Чтобы не повторять все кейсы, тестируемые в самой библиотеке, можно попытаться запускать родные тесты — https://github.com/enb-make/enb-stylus/pull/61.

> Ещё, кажется, я не очень удачно назвал свои технологии. В частности имя моей технологии совпадает с именем подключаемой библиотеки.

Тут я врядли подскажу =)
